### DISCUSSION TITLE
Use expressions in behavior properties or variable initial values

### DISCUSSION THREAD

**Discussion by PascalLadalle** [4ian/GDevelop#2342] [Enhancement ideas (and their technical discussions)]
I recommended my TimedBackAndForth extension to a user, and he said he wanted to use a **random** speed.
It's not possible to type in math in a number field, and it's not possible to change the property through events (because it's a lifecycle method?)
I edited the extension, converted the Speed property to a string, edited the events to convert the property to a number, and it works fine if I enter 200 as string value, but it doesn't work when I enter ToString(Random(200)).
I tried to convert and display the property, likeso:
![image](https://user-images.githubusercontent.com/25978830/77860019-07400980-720d-11ea-96bc-340e4ae98aae.png)

And I got:
![image](https://user-images.githubusercontent.com/25978830/77859886-3b66fa80-720c-11ea-9e0d-9ed33fc80093.png)

I tried moving the problem to a variable, likeso:
![image](https://user-images.githubusercontent.com/25978830/77860272-a285ae80-720e-11ea-8eef-47db974d24a0.png)
But it's not working either.

Shouldn't variables and math functions be working in these fields?

**Comment by PascalLadalle** [comment]
I was offered a JS solution by Erigatohishima:
![image](https://user-images.githubusercontent.com/25978830/77860866-e0380680-7211-11ea-90c8-4f27c5c6e86d.png)
But I don't think JS should be required for that.

**Comment by 4ian** [comment]
Properties are primitive types - they are not expressions. This can be seen in the UI because there is no expression editor to edit them - they are just a value.

> and it works fine if I enter 200 as string value, but it doesn't work when I enter ToString(Random(200)).

It's expected.

> I was offered a JS solution by Erigatohishima:

Never use eval! I don't think JavaScript is even useful for that. eval is dangerous and is a nightmare for performance. Don't use it. JavaScript has functions to get a random value. GDevelop has one too. It is more than enough.

Back to the original problem:

> he said he wanted to use a random speed.

Is the behavior supporting to customize the speed with an action? If yes, why not using an action to set the speed to a Random value?

Second solution, not involving any extra events in the scene events sheet: you edit the behavior, add a "onCreated" behavior function and inside it you set the property to a random value.

> Shouldn't variables and math functions be working in these fields?

Not necessarily. This requires very careful thinking. There is a high risk of "order of initialization fiasco". What if you try to read the value of a variable inside the behavior... but the variable itself has its value initialized from another variable... which itself has a value initalized from the previous variable?

We have two issues:
- If the engine initialize the behavior BEFORE the variable, you're in a mess because the variable does not exist yet. So you're getting zero (the value when a variable is not set). And this is a nightmare to debug because you don't understand and nothing can help you finding the issue.
- If the engine initialize a variable before another variable, or do the reverse later, you're getting DIFFERENT results. Again a nightmare to debug. "Why is my game changing everytime I run it?"

This being said, I understand the idea and that's maybe something we could add but with severely limited options, like only "pure" expressions like mathematical expressions, random ones and a few others. But in the meantime, better be explicit and just set your property using an action - if the behavior allow you to do so.

**Comment by PascalLadalle** [comment]
At least basic math like random would be nice, yes.
Thanks for the solution and explanations!
